### PR TITLE
Implement render mode `fog_disabled` and BaseMaterial3D setting Disable Fog

### LINKED ISSUE #56374: Fog volume breaks additive unshaded materials
### Godot version

4.0 dev

### System information

Arch Linux, KDE Plasma, Radeon RX580, amdgpu driver

### Issue description

The plane is textured with a radial gradient and should produce a smooth transition between nothing and white. The fog volume however breaks that and makes edges of the mesh clearly visible, which is not what the mesh is supposed to look like.

1: this is the expected look
2: this is the unexpected look when fog volume interacts with the additive unshaded mesh 

image

### Steps to reproduce

1. Create an unshaded additive material with a gradient from black to color.
2. Add a fog volume and enable fog volume in Environment settings
3. Make the fog volume overlap the mesh
4. See as the mesh looses it's indented look

### Minimal reproduction project

[Fog Volume Additive.zip](https://github.com/godotengine/godot/files/7796413/Fog.Volume.Additive.zip)

### PR DESCRIPTION
**Pull request by lemilonkh**
This fixes #56374 (and its duplicates like #69268).
A new render mode called `fog_disabled` is added. It can be used in spatial shaders (using `render_mode fog_disabled`) and from `StandardMaterial3D` using the new toggle `Disable Fog` found in the `Shading` section.

This works for the Forward+, Mobile and Compatibility rendering backends.

Documentation for the new render mode and material setting has been added in this PR: godotengine/godot-docs#7875

## Screenshots
Disable Fog setting enabled |  Disable Fog setting disabled
:----------------------------------------:|:-----------------------------------------:
image | image
image | image

## Minimal reproduction project
The minimal reproduction project from #56374 has been modified to include a DirectionalLight3D (without which the volumetric fog wouldn't render anymore), and the new `disable_fog` property has been enabled on the unshaded additively blended mesh. I've also enabled the depth-based fog to test the effect on the Mobile and Compatibility render backends as they don't support volumetric fog, and added a cube with a shader material using `render_mode unshaded;` to demonstrate this working as well.

It can be downloaded here:
[Fog_Volume_Additive.zip](https://github.com/godotengine/godot/files/12507940/Fog_Volume_Additive.zip)

### CODE REVIEW

**Comment by AThousandShips** [comment]
The flag needs to be documented, see the CI errors

**Comment by lemilonkh** [comment]
I'm not quite sure what causes the failure in this pipeline: https://github.com/godotengine/godot/actions/runs/6067098408/job/16458879823?pr=81286
It seems to be related to freeing an invalid key in the the ClassDB?

**Comment by AThousandShips** [comment]
Unrelated to this PR, sporadic error

File: doc/classes/BaseMaterial3D.xml [hunk 1]
```diff
@@ -697,6 +700,9 @@
 		<constant name="FLAG_ALBEDO_TEXTURE_MSDF" value="20" enum="Flags">
 			Enables multichannel signed distance field rendering shader.
 		</constant>
+		<constant name="FLAG_DISABLE_FOG" value="21" enum="Flags">
+			Disables receiving depth-based or volumetric fog.
++		</constant>
```
Comment (AThousandShips) [inline doc/classes/BaseMaterial3D.xml:705]:
```suggestion
		</constant>
```

**Review by clayjohn — APPROVED** [review]
Code looks good to me and I have tested locally and everything seems to be working fine!

Great work!

For maintainers, this should be merged in 4.2, but not cherrypicked to 4.1

**Comment by YuriSizov** [comment]
Could you please squash your commits into one? Make sure that the final commit has a short but descriptive message (the title of this PR is a good option). See [this documentation](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase), if you need help with squashing.

**Comment by lemilonkh** [comment]
Done :sparkles:

**Comment by YuriSizov** [comment]
Thanks, and congrats on your first merged Godot PR!

**Comment by EzraT** [comment]
Thank you!

**Pull request merged**